FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+         * from having more than 32 publishes. If many publishes are desired, semaphores
+         * can be used instead of task notifications. */
+        pxContexts[ i ]->notificationBit = 1U << i;
+        payloadBuffers[ i ] = ( char * ) pvPortMalloc( DYNAMIC_BUFFER_SIZE );
+        topicBuffers[ i ] = ( char * ) pvPortMalloc( DYNAMIC_BUFFER_SIZE );
+        snprintf( payloadBuffers[ i ], DYNAMIC_BUFFER_SIZE, PUBLISH_PAYLOAD_FORMAT, i + 1 );
+        snprintf( topicBuffers[ i ], DYNAMIC_BUFFER_SIZE, PUBLISH_TOPIC_FORMAT_STRING, i + 1 );
+        /* Set publish info. */
+        memset( ( void * ) &( pxPublishes[ i ] ), 0x00, sizeof( MQTTPublishInfo_t ) );
+        pxPublishes[ i ].pPayload = payloadBuffers[ i ];
+        pxPublishes[ i ].payloadLength = strlen( payloadBuffers[ i ] );
+        pxPublishes[ i ].pTopicName = topicBuffers[ i ];
+        pxPublishes[ i ].topicNameLength = ( uint16_t ) strlen( topicBuffers[ i ] );
+        pxPublishes[ i ].qos = MQTTQoS2;
+        pxContexts[ i ]->pPublishInfo = &( pxPublishes[ i ] );
+        LogInfo( ( "Adding publish operation for message %s \non topic %.*s\n",
+                   payloadBuffers[ i ],
+                   pxPublishes[ i ].topicNameLength,
+                   pxPublishes[ i ].pTopicName ) );
+        prvCreateCommand( PUBLISH, pxContexts[ i ], prvCommandCallback, &xCommand );
+        prvAddCommandToQueue( &xCommand );
+        LogInfo( ( "Publish operation queued. Sleeping for %d ms.\n", PUBLISH_DELAY_ASYNC_MS ) );
+        vTaskDelay( pdMS_TO_TICKS( PUBLISH_DELAY_ASYNC_MS ) );
+    }
+
+    LogInfo( ( "Finished publishing\n" ) );
+
+    for( int i = 0; i < PUBLISH_COUNT; i++ )
+    {
+        if( pxContexts[ i ] == NULL )
+        {
+            /* Don't try to free anything that wasn't initialized. */
+            continue;
+        }
+
+        LogInfo( ( "Waiting to free publish context %d.", i ) );
+        xTaskNotifyWait( 0, ( 1U << i ), &ulNotification, DEMO_TICKS_TO_WAIT );
+        configASSERT( ( ulNotification & ( 1U << i ) ) == ( 1U << i ) );
+        prvDestroyCommandContext( pxContexts[ i ] );
+        vPortFree( pxContexts[ i ] );
+        vPortFree( topicBuffers[ i ] );
+        vPortFree( payloadBuffers[ i ] );
+        LogInfo( ( "Publish context %d freed.", i ) );
+        pxContexts[ i ] = NULL;
+    }
+
+    /* Notify main task this task can be deleted. */
+    xTaskNotify( xMainTask, TASK1_COMPLETE_BIT, eSetBits );
+}
+
+/*-----------------------------------------------------------*/
+
+void prvSubscribeTask( void * pvParameters )
+{
+    ( void ) pvParameters;
+    MQTTSubscribeInfo_t xSubscribeInfo;
+    Command_t xCommand;
+    MQTTPublishInfo_t * pxReceivedPublish = NULL;
+    static uint16_t usNumReceived = 0;
+    uint32_t ulNotification;
+    CommandContext_t xContext;
+    PublishElement_t xReceivedPublish;
+    uint16_t usWaitCounter = 0;
+
+    xSubscribeInfo.qos = MQTTQoS0;
```

Comment:
Is there a reason why QoS0 was chosen for subscribe where as publishes are QoS2?

Reply:
No reason for the QoS 0, besides that QoS 0 is simplest and I didn't see a point in using QoS 2 if it wasn't necessary. QoS 2 was chosen for the publishes since I wanted to show an operation that waited for an ack packet, and wouldn't complete immediately like a QoS 0 publish. I could have also used QoS 1.

Reply:
Thanks for letting me the reason for this decision. A comment would just be fine.